### PR TITLE
agent: add support for ttrpc

### PR DESCRIPTION
**Pull request by Xynnn007**
Using ttrpc instead grpc feature of image-rs for kata-agent

Close https://github.com/kata-containers/kata-containers/issues/6219

cc @arronwy @jiangliu

### CODE REVIEW

File: src/agent/Cargo.lock
Comment (c3d) [inline src/agent/Cargo.lock:91]: Thanks @Xynnn007. Would it be possible to split your `Cargo.lock` changes into two parts, one that derives from build updates (which arguably you don't even need to commit) and the part that specifically adds the required `ttrpc` support? That would make it easier to review.

File: src/agent/Cargo.toml
Comment (c3d) [inline src/agent/Cargo.toml:74]: Do you really want to go to a hash-based version here?
Also, I find the feature changes confusing. Why drop kata-cc-s390x?

File: src/agent/Cargo.toml
Comment (c3d) [inline src/agent/Cargo.toml:77]: Same comment as above.

**Review by c3d — COMMENTED** [review]
Thanks @Xynnn007. I have a few questions about the packaging of this PR.

File: src/agent/Cargo.toml
Comment (Xynnn007) [reply src/agent/Cargo.toml:74]: This is because now the `"kata-cc-s390"` feature equals to `["encryption", "keywrap-grpc", "snapshot-overlayfs", "signature-simple", "getresource"]`, where `"keywrap-grpc"` is enabled already.  `"keywrap-grpc"` and `"keywrap-ttrpc"` are inclusive features now.

I understand your concern. Maybe we can have a fix in `image-rs` to replace `"keywrap-grpc"` with `"keywrap-ttrpc"` in `"kata-cc-s390"`

File: src/agent/Cargo.toml
Comment (Xynnn007) [reply src/agent/Cargo.toml:77]: Same as above

File: src/agent/Cargo.toml
Comment (c3d) [reply src/agent/Cargo.toml:74]: Hmm. That seems confusing.
The s390x platform is somewhat different, and I find it much clearer to have something that tells us explicitly "this is s390x-aware" than having the details of how it's done. I think the idea is to contrast that to regular `kata-cc`, right?

Comment (Xynnn007) [reply src/agent/Cargo.toml:74]: This confusing `s390x` feature origins from this issue https://github.com/kata-containers/kata-containers/issues/5582  To put it simply this feature only controls whether `signature-cosign` is enabled. To make it transparent for kata-agent, a `s390x` feature is used which performs as a temporary solusion.

BTW, I'm working on the chain of dependencies of cosign to make it built-able on `s390x`. In future we can combine the two different features into one

File: src/agent/Cargo.lock
Comment (Xynnn007) [reply src/agent/Cargo.lock:91]: Great suggestion! Let me fix this and then force push with other change altogether

File: src/agent/Cargo.toml
Comment (c3d) [reply src/agent/Cargo.toml:74]: I'm afraid you misunderstood what confused me. I find 's390x' to be relatively clear, because it expresses an intent, it answers a "why?" question. By contrast, elaborating the list of sub-features hides that intent, it answers a "what?" question nobody really cares about.

To illustrate why I find this confusing, consider if I replaced:
```
   async fn pull_image(&self, req: &image::PullImageRequest) -> Result<String> {
```

with

```
   async fn set_env_var_then_setup_https_proxy_then_build_cid_from_request_then_unpack_pause_image_if_necessary_otherwise_fetch_aa_kbc_params_and_so_on(&self, req: &image::PullImageRequest) -> Result<String> {
```

I suspect you would agree that the second name is confusing, no? 😄

File: src/agent/src/image_rpc.rs
Comment (c3d) [inline src/agent/src/image_rpc.rs:32]: Shouldn't those be renamed to `AA_KEYPROVIDER_URI` and `AA_GETRESOURCE_URI`?

File: src/agent/Cargo.toml
Comment (Xynnn007) [reply src/agent/Cargo.toml:74]: Aha, you're right. Sorry for not explaining myself clearly. :(

> Maybe we can have a fix in `image-rs` to replace `"keywrap-grpc"` with `"keywrap-ttrpc"` in `"kata-cc-s390"`

1. A pr in image-rs should be patched first to include enough "nobody-really-cares-about" features for `"kata-cc-s390x"` and `"kata-cc"`
2. Then let me continue on this PR, use previous `"kata-cc-s390x"` and `"kata-cc"`
3. When no need to distinguish `"kata-cc-s390x"` and `"kata-cc"` in image-rs, we will delete feature `"kata-cc-s390x"` in image-rs and then in `kata-agent`

Is this clear and does it make sense? :P

File: src/agent/src/image_rpc.rs
Comment (Xynnn007) [reply src/agent/src/image_rpc.rs:32]: Oh, yes. Sorry for this. I'll fix this

File: src/agent/Cargo.toml
Comment (c3d) [reply src/agent/Cargo.toml:74]: Yes, it makes total sense. I don't think you need a separate PR. This could be preparatory commits in this PR, if that's easier.

Comment (c3d) [reply src/agent/Cargo.toml:74]: Or it could be a separate PR, whatever is easier for you.

File: src/agent/src/image_rpc.rs
Comment (jepio) [inline src/agent/src/image_rpc.rs:33]: it would make more sense to keep these domain sockets in `/run/...`

Comment (Xynnn007) [reply src/agent/src/image_rpc.rs:33]: Yes, you're right! Thanks. We'll fix this from AA side, and then will update this PR

Comment (jepio) [reply src/agent/src/image_rpc.rs:33]: Is this hardcoded? Or can we pass the socket path as a command line argument when starting AA

Comment (Xynnn007) [reply src/agent/src/image_rpc.rs:33]: This is a default value in attestation-agent, and surely can pass from command line. For kata-agent, using default value is enough I think

**Comment by Xynnn007** [comment]
Hi @c3d now the `Cargo.lock` is changed by `cargo update -p async-trait` because if use the upstream-version cargo.lock
```
root@xynnn:/home/kata-containers/src/agent# cargo build
    Updating git repository `https://github.com/confidential-containers/image-rs`
    Updating git repository `https://github.com/confidential-containers/ocicrypt-rs.git`
    Updating crates.io index
error: failed to select a version for `async-trait`.
    ... required by package `ocicrypt-rs v0.1.0 (https://github.com/confidential-containers/ocicrypt-rs.git?rev=b720152#b7201528)`
    ... which satisfies git dependency `ocicrypt-rs` of package `image-rs v0.1.0 (https://github.com/confidential-containers/image-rs?rev=e7f57a4#e7f57a47)`
    ... which satisfies git dependency `image-rs` of package `kata-agent v0.1.0 (/home/kata-containers/src/agent)`
versions that meet the requirements `^0.1.61` are: 0.1.66, 0.1.65, 0.1.64, 0.1.63, 0.1.62, 0.1.61

all possible versions conflict with previously selected packages.

  previously selected package `async-trait v0.1.60`
    ... which satisfies dependency `async-trait = "^0.1"` (locked to 0.1.60) of package `cached v0.38.0`
    ... which satisfies dependency `cached = "^0.38.0"` (locked to 0.38.0) of package `sigstore v0.3.3`
    ... which satisfies dependency `sigstore = "^0.3.0"` (locked to 0.3.3) of package `image-rs v0.1.0 (https://github.com/confidential-containers/image-rs?rev=e7f57a4#e7f57a47)`
    ... which satisfies git dependency `image-rs` of package `kata-agent v0.1.0 (/home/kata-containers/src/agent)`

failed to select a version for `async-trait` which could resolve this conflict
```

The change is from
```
root@xynnn:/home/kata-containers/src/agent# cargo update -p async-trait
    Updating crates.io index
    Updating git repository `https://github.com/confidential-containers/image-rs`
    Updating git repository `https://github.com/confidential-containers/ocicrypt-rs.git`
    Removing async-stream v0.3.3
    Removing async-stream-impl v0.3.3
    Updating async-trait v0.1.60 -> v0.1.66
    Removing axum v0.6.1
    Removing axum-core v0.3.0
    Removing http-range-header v0.3.0
    Removing hyper-timeout v0.4.1
      Adding image-rs v0.1.0 (https://github.com/confidential-containers/image-rs?rev=e7f57a4#e7f57a47)
    Removing image-rs v0.1.0 (https://github.com/confidential-containers/image-rs?rev=v0.4.0#948db858)
    Removing matchit v0.7.0
    Removing ocicrypt-rs v0.1.0 (https://github.com/confidential-containers/ocicrypt-rs?rev=453c699#453c699a)
      Adding ocicrypt-rs v0.1.0 (https://github.com/confidential-containers/ocicrypt-rs.git?rev=b720152#b7201528)
    Removing prettyplease v0.1.22
    Removing prost v0.11.5
    Removing prost-build v0.11.5
    Removing prost-derive v0.11.5
    Removing prost-types v0.11.5
      Adding protobuf v3.2.0
      Adding protobuf-codegen v3.2.0
      Adding protobuf-parse v3.2.0
      Adding protobuf-support v3.2.0
    Removing sync_wrapper v0.1.1
    Removing tokio-io-timeout v1.2.0
    Removing tonic v0.8.3
    Removing tonic-build v0.8.4
    Removing tower v0.4.13
    Removing tower-http v0.3.5
    Removing tower-layer v0.3.2
    Removing tracing-futures v0.2.5
      Adding ttrpc v0.7.1
      Adding ttrpc-codegen v0.4.1
      Adding ttrpc-compiler v0.6.1
```

Please check if this PR looks good now : ), cc @jepio. 


Also, this PR will bring KBS Resource URI scheme from `image-rs`. The test encrypted images used now will not be decrypted after merging. We need to make new test images. cc @stevenhorsman

File: src/agent/Cargo.toml
Comment (jepio) [inline src/agent/Cargo.toml:74]: can you use a longer hash (either full or atleast 12 chars) so that we don't have problems with non-unique references later

File: src/agent/Cargo.toml
Comment (jepio) [inline src/agent/Cargo.toml:77]: here too

**Review by jepio — APPROVED** [review]
LGTM (once the `rev` is extended)

File: src/agent/Cargo.toml
Comment (Xynnn007) [reply src/agent/Cargo.toml:74]: Ok!

**Comment by Xynnn007** [comment]
A quick question: is kata-cc allowed to use a non-tagged version of `image-rs`? Or we should wait until the next release of `image-rs` and then continue the PR? Sorry I did not notice this before : ( @c3d @jepio

**Comment by stevenhorsman** [comment]
> A quick question: is kata-cc allowed to use a non-tagged version of `image-rs`? Or we should wait until the next release of `image-rs` and then continue the PR? Sorry I did not notice this before : ( @c3d @jepio

@Xynnn007 Yes - we can definitely take non-tagged versions of components between releases and we strongly encourage it, so that we don't only test the integration of updated features during release creation, so thank you for raising this PR nice and early in the release!

Specifically in this case if there is a breaking change and we need to update the test images then we'd need to wait for the related tests change to be ready before we can merge it, but the integration tests should fail and stop that happening anyway.

Is there a kata-containers tests issue covering this change with any links to the docs need to set up the new images in?

**Comment by stevenhorsman** [comment]
The commit message check is failing as it doesn't match the kata [patch-format rules](https://github.com/kata-containers/community/blob/main/CONTRIBUTING.md#patch-format)

I think adding `Fixes: #6219` should help.

**Comment by Xynnn007** [comment]
Hey @stevenhorsman thanks a lot for the explaination! I'm quite not familiar with this :D . Let me update the image-rs rev and follow your guide!

> Is there a kata-containers tests issue covering this change with any links to the docs need to set up the new images in?

Let me open a new issue for this

**Comment by Xynnn007** [comment]
FYI, the new rev is not the same as the old one, as the new one fixed a bug recently.

File: src/agent/Cargo.toml
Comment (Xynnn007) [reply src/agent/Cargo.toml]: sample signing is now included in the big feature `kata-cc` now. Here https://github.com/confidential-containers/image-rs/blob/main/Cargo.toml#L75

**Comment by arronwy** [comment]
@Xynnn007 we also need upgrade `attestation-agent` with ttrpc support： https://github.com/kata-containers/kata-containers/blob/CCv0/versions.yaml#L194

File: src/agent/Cargo.toml
Comment (jepio) [inline src/agent/Cargo.toml:80]: Why is this necessary? The commit message should include this information.

Comment (Xynnn007) [reply src/agent/Cargo.toml:80]: Yes, added it.

Comment (Xynnn007) [reply src/agent/Cargo.toml:80]: When the same patch is applied on `image-rs`, it will not work for the higher callers, like `kata-agent`

Comment (jepio) [reply src/agent/Cargo.toml:80]: This is unfortunate, but i assume it's temporary, so ok

Comment (Xynnn007) [reply src/agent/Cargo.toml:80]: Yes. When the new version of `oci-distribution` is published on the `crates.io`, we can get rid of this

**Review by arronwy — APPROVED** [review]
Thanks @Xynnn007 , LGTM!

File: versions.yaml
Comment (jepio) [inline versions.yaml:194]: this needs to be a full-length hash, it doesn't work otherwise.

Comment (Xynnn007) [reply versions.yaml]: Thanks, I've fixed it.

**Comment by jepio** [comment]
Tests are failing with an attestation agent panic:
```
called `Result::unwrap()` on an `Err` value: invalid socket address syntax', src/main.rs:31:37
```

**Comment by Xynnn007** [comment]
> Tests are failing with an attestation agent panic:
> 
> ```
> called `Result::unwrap()` on an `Err` value: invalid socket address syntax', src/main.rs:31:37
> ```

Oh sorry, , I've update the `rootfs.sh` to enable ttrpc by default and rebase the latest version of branch CCv0.

**Comment by jepio** [comment]
Now failing with:
```
[2023-03-30T08:02:00Z ERROR attestation_agent::rpc::getresource::ttrpc] Call AA-KBC to get resource failed: Received unknown resource name: default/credential/test
{\"msg\":\"image_client.config.auth set to: true\",\"level\":\"INFO\",\"ts\":\"2023-03-30T08:02:00.292965207Z\",\"version\":\"0.1.0\",\"subsystem\":\"rpc\",\"source\":\"agent\",\"name\":\"kata-agent\",\"pid\":\"121\"}
{\"msg\":\"enable_signature_verification set to: true\",\"level\":\"INFO\",\"ts\":\"2023-03-30T08:02:00.293025547Z\",\"subsystem\":\"rpc\",\"pid\":\"121\",\"version\":\"0.1.0\",\"source\":\"agent\",\"name\":\"kata-agent\"}
{\"msg\":\"pull image \"confidential-containers_signed_0\", bundle path \"/run/kata-containers/confidential-containers_signed_0\"\",\"level\":\"INFO\",\"ts\":\"2023-03-30T08:02:00.293085751Z\",\"version\":\"0.1.0\",\"pid\":\"121\",\"name\":\"kata-agent\",\"subsystem\":\"rpc\",\"source\":\"agent\"}
[2023-03-30T08:02:00Z ERROR attestation_agent::rpc::getresource::ttrpc] Call AA-KBC to get resource failed: Received unknown resource name: default/security-policy/test
{\"msg\":\"pull and unpack image \"quay.io/kata-containers/confidential-containers:signed\", cid: \"confidential-containers_signed_0\", with image-rs failed with \"Security validate failed: ttrpc request error

Caused by:
    rpc status: Status { code: INTERNAL, message: \"[ERROR:attestation-agent] AA-KBC get resource failed: Received unknown resource name: default/security-policy/test\", details: [], special_fields: SpecialFields { unknown_fields: UnknownFields { fields: None }, cached_size: CachedSize { size: 0 } } }\". \",\"level\":\"ERRO\",\"ts\":\"2023-03-30T08:02:00.946668452Z\",\"subsystem\":\"rpc\",\"pid\":\"121\",\"source\":\"agent\",\"version\":\"0.1.0\",\"name\":\"kata-agent\"}
```

**Comment by jepio** [comment]
@Xynnn007 latest failure is:
```
# Mar 31 12:15:13 ubuntu20-4773f0 kata[96128]: time="2023-03-31T12:15:13.289176106Z" level=debug msg="reading guest console" console-protocol=unix console-url=/run/vc/vm/e8667d777f5855634b3731855964e4c41c1fe6239bd0cd6cfac3ea78ff1dec4a/console.sock name=containerd-shim-v2 pid=96128 sandbox=e8667d777f5855634b3731855964e4c41c1fe6239bd0cd6cfac3ea78ff1dec4a source=virtcontainers subsystem=sandbox vmconsole="{\"msg\":\"pull and unpack image \\\"quay.io/kata-containers/confidential-containers:cosign-signed\\\", cid: \\\"confidential-containers_cosign-signed_0\\\", with image-rs failed with \\\"Security validate failed: Validate image failed: No such file or directory (os error 2)\\\". \",\"level\":\"ERRO\",\"ts\":\"2023-03-31T12:15:13.276058764Z\",\"subsystem\":\"rpc\",\"version\":\"0.1.0\",\"source\":\"agent\",\"pid\":\"121\",\"name\":\"kata-agent\"}"
# Mar 31 12:15:13 ubuntu20-4773f0 kata[96128]: time="2023-03-31T12:15:13.289253908Z" level=error msg="agent pull image err. rpc error: code = Internal desc = Security validate failed: Validate image failed: No such file or directory (os error 2)" name=containerd-shim-v2 pid=96128 sandbox=e8667d777f5855634b3731855964e4c41c1fe6239bd0cd6cfac3ea78ff1dec4a source=virtcontainers subsystem=kata_agent
# Mar 31 12:15:13 ubuntu20-4773f0 kata[96128]: time="2023-03-31T12:15:13.289346209Z" level=error msg="kata runtime PullImage err. rpc error: code = Internal desc = Security validate failed: Validate image failed: No such file or directory (os error 2)" name=containerd-shim-v2 pid=96128 sandbox=e8667d777f5855634b3731855964e4c41c1fe6239bd0cd6cfac3ea78ff1dec4a source=containerd-kata-shim-v2
```
full log here: http://jenkins.katacontainers.io/job/kata-containers-CCv0-ubuntu-20.04-x86_64-CC_CRI_CONTAINERD-PR/695/consoleText
This is the cosign test case

**Comment by jepio** [comment]
Merge needs to be synced with other PRs.

**Pull request merged**
